### PR TITLE
Enable option for subword regularization in more tokenizers.

### PR DESCRIPTION
**Pull request by PhilipMay**
see https://github.com/huggingface/transformers/pull/11149#pullrequestreview-643686428

### `AlbertTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `BarthezTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `BertGenerationTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `BigBirdTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `CamembertTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `DebertaV2Tokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `M2M100Tokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `MarianTokenizer` - has src and target tokenizer
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `MBart50Tokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `PegasusTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `ReformerTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `Speech2TextTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `T5Tokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `XLMProphetNetTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] check
- [x] refactor test to follow DRY
- <s>remove obscure function argument called `sample`</s>
### `XLNetTokenizer`
- [x] add `sp_model_kwargs` param with test
- [x] add pickle support with test
- [x] remove obscure function argument called `sample`
- [x] check
- [x] refactor test to follow DRY
### `XML RoBERTa`
- [x] refactor test to follow DRY
### General
- [x] check if we changed all tokenizers
- [x] add typing
- [x] check if tok. is used in other functions
- [x] also add changes to XLM RoBERTa tokenizer
### After review
- [x] fix type comments with default `None`
- [x] possibly remove `test_sentencepiece_skip_back_convert_check`

### CODE REVIEW

**Comment by PhilipMay** [comment]
I found this somehow obscure function argument called `sample` at `AlbertTokenizer`:

https://github.com/huggingface/transformers/blob/52166f672ed337934d90cc525c226d93209e0923/src/transformers/models/albert/tokenization_albert.py#L189

It seems to enable subword regularization but with fixed parameters for `nbest_size` and `alpha`.

https://github.com/google/sentencepiece/blob/351600c2971401f4e849147579aa1b5d42f614e1/python/src/sentencepiece/__init__.py#L110-L111

I would remove that `sample` parameter and replace that with my solution which is more flexible. But that would mean we have a breaking change. As an alternative I could add my solution but keep the `sample` argument. 

What do you think? @sgugger @LysandreJik @stefan-it

**Comment by sgugger** [comment]
This argument is not called from anywhere so it's only accessible if users somehow rewrote the tokenize method to pass it along to the private method `_tokenize`. Therefore I think it's fine to do the breaking change and clean up the code using `sample=True`, but let's see what @patrickvonplaten and @LysandreJik think before going forward (note that Lysandre is on vacation until this Wednesday so he'll reply at the end of the week :-) ).

**Comment by LysandreJik** [comment]
Yes, removing the `sample` and cleaning up the `_tokenize()` method sounds good to me. As @sgugger said, it is private and nowhere is a `sample` or a `**kwargs` passed to that method.

**Comment by PhilipMay** [comment]
rebase upstream/master done

**Review by LysandreJik — APPROVED** [review]
Yes, LGTM! Thanks a lot.

**Comment by PhilipMay** [comment]
> Yes, LGTM! Thanks a lot.

Hey @LysandreJik - this is not done yet. Please do not merge now. ;-)

**Comment by LysandreJik** [comment]
Oh, I was misled! There are indeed a few tokenizers remaining. Thank you for letting me know!

**Comment by PhilipMay** [comment]
This is ready to be merged from my point of view.

**Comment by sgugger** [comment]
Can you take care of the merge conflicts? Will review tomorrow :-)

**Comment by PhilipMay** [comment]
> Can you take care of the merge conflicts? Will review tomorrow :-)

@sgugger All conflicts resolved & green CI

File: src/transformers/models/albert/tokenization_albert.py [hunk 1]
```diff
@@ -102,6 +102,20 @@ class AlbertTokenizer(PreTrainedTokenizer):
         mask_token (:obj:`str`, `optional`, defaults to :obj:`"[MASK]"`):
             The token used for masking values. This is the token used when training this model with masked language
             modeling. This is the token which the model will try to predict.
+        sp_model_kwargs (:obj:`dict`, `optional`, defaults to :obj:`None`):
```
Comment (sgugger) [inline src/transformers/models/albert/tokenization_albert.py:105]:
```suggestion
        sp_model_kwargs (:obj:`dict`, `optional`):
```
We don't specify the default when it's None.

I'm putting it once here, but it applies to every other docstring as well.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

For the tests, I see a lot of repetition so I wonder if it would be possible to have the two tests be common tests (with a class attribute `test_subword_regularization` in the Tester False by default that the classes where we want tot test would set to True). I think it would also be cleaner to have the kwargs passed to `get_tokenizer` method so you can use:
```
tokenizer = self.get_tokenizer(keep_accents=True, sp_model_kwargs={"enable_sampling": True, "alpha": 0.1, "nbest_size": -1})
```
in your common test.
```

**Review by LysandreJik — APPROVED** [review]
Cool, thanks a lot for going through all of those! 

Great work on the tests, this is great. The tests could indeed be refactored in a common test if you feel like it.

File: src/transformers/models/albert/tokenization_albert.py (hunk 1 above)
Comment (PhilipMay) [reply src/transformers/models/albert/tokenization_albert.py]: I changed this and created a new Issue as a reminder to apply the same change on other locations: https://github.com/huggingface/transformers/issues/11687

**Comment by PhilipMay** [comment]
This is strange:

`FAILED tests/test_hf_api.py::HfApiEndpointsTest::test_list_repos_objs - reque...`

Will trigget CI again,,,

**Comment by PhilipMay** [comment]
@LysandreJik @sgugger Tests are refactored and DRY now. CI is green again.
IMO ready for merge.

Maybe you want to investigate the flaky test (see my comment above).

**Review by sgugger — APPROVED** [review]
The refactor looks good to me, thanks a lot!

**Review by LysandreJik — APPROVED** [review]
Fantastic, thanks a lot @PhilipMay! Very clean PR.

**Pull request merged**